Document/0x05h-Testing-Platform-Interaction.md

```diff
@@ -1587,6 +1587,33 @@ In order to test for proper updating: try downloading an older version of the ap
 Next, verify whether or not you can continue to use the application without updating it. If an update prompt is given, verify if you can still use the application by canceling the prompt or otherwise circumventing it through normal application usage. This includes validating whether the backend will stop calls to vulnerable backends and/or whether the vulnerable app-version itself is blocked by the backend.
 Lastly, see if you can play with the version number of a man-in-the-middled app and see how the backend responds to this (and if it is recorded at all for instance).
 
+## Testing App Notifications (MSTG-PLATFORM-12)
+
+### Overview
+
+Android allows for applications to create [notifications](https://developer.android.com/guide/topics/ui/notifiers/notifications "Notifications Overview"), a form of message that is displayed by the Android system outside of the application's UI. This is useful for applications to display small messages to the user while they use other applications on their device.
+
+It is important to understand that notifications should never be considered private. When a notification is handled by the Android system it is broadcasted system-wide and any application running with a [NotificationListenerService](https://developer.android.com/reference/kotlin/android/service/notification/NotificationListenerService "https://developer.android.com/reference/kotlin/android/service/notification/NotificationListenerService") can listen for these notifications to receive them in full. From here the Listener application can handle these notifications however it wants.
+
+There are many known malware examples such as Joker, and Terracotta which abuse the `NotificationListenerService` to listen for notifications on the device and then send them to attacker-controlled C2 infrastructure. Furthermore there are a number of apps on the Google Play Store that provide notification logging; locally logging any notifications on the Android system. This highlights that notifications are in no way private on Android.  
+
+Consider a hypothetical application that features a two-factor authentication (2FA) function built into the application. When a user goes to sign in to the application on a web browser they are prompted for a 2FA code from their mobile device, the application creates a notification that contains this code for ease of access.
+
+A malicious application listening for notifications could obtain this 2FA code from the notification and send it to the attacker for use. The application should instead use the notification to prompt the user to go into the application its self to obtain the 2fa code.
```

Comment:
What about apps like GitHub/Twitter/etc. If you login in the browser or in the app you'll get an SMS with the 2FA code, so any app will be able to see it. What should be done on that case?

Reply:
On re-reading that first paragraph I can see why you'd be confused; made sense in my head at the time. 

Basically some applications I've tested in the past have had in built 2fa functions (Steam guard does this for example) so that when you sign into the account on a web browser a notification pops up on your phone telling you your 2fa. A malicious app can of course listen for notifications and grab this notification; thus its not just limited to 2fa on SMS.


To answer your other question 2FA over SMS is in my opinion (and the [opinion ](https://blog.sucuri.net/2020/01/why-2fa-sms-is-a-bad-idea.html )[of ](https://www.kaspersky.com/blog/2fa-practical-guide/24219/ )[many ](https://www.theverge.com/2017/9/18/16328172/sms-two-factor-authentication-hack-password-bitcoin)[others](https://www.cnet.com/how-to/do-you-use-sms-for-two-factor-authentication-heres-why-you-shouldnt/)) a weak form of 2FA for reasons both described here and inherent to SMS. 

But to focus on the Android system side and to answer what should be done in this case: Dont use sms 2fa or 2fa that generates popups.

Ultimately this is a weakness of the android system and how it handles notifications. Notifications are not private like the per app notifications of iOS and thus the best thing to be done is to avoid generating notifications.

I hope this makes sense and I'm happy to elaborate more if need be, ill also do some modifications to this section.